File: ios/Classes/Crypto/Crypto.swift
//
//  Scrypt.swift
//  Runner
//
//  Created by 蒋治国 on 2022/7/8.
//

import Nkn

class Crypto : ChannelBase, FlutterStreamHandler {

    static var instance: Crypto = Crypto()
    let qryptoQueue = DispatchQueue(label: "org.nkn.mobile/native/crypto/queue", qos: .default, attributes: .concurrent)
    private var qryptoWorkItem: DispatchWorkItem?

    var methodChannel: FlutterMethodChannel?
    let METHOD_CHANNEL_NAME = "org.nkn.mobile/native/crypto_method"
    var eventSink: FlutterEventSink?
    
    public static func register(controller: FlutterViewController) {
        instance.install(binaryMessenger: controller as! FlutterBinaryMessenger)
    }

    func install(binaryMessenger: FlutterBinaryMessenger) {
        self.methodChannel = FlutterMethodChannel(name: METHOD_CHANNEL_NAME, binaryMessenger: binaryMessenger)
        self.methodChannel?.setMethodCallHandler(handle)
    }

    func uninstall() {
        self.methodChannel?.setMethodCallHandler(nil)
    }

    func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        eventSink = events
        return nil
    }

    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        // eventSink = nil
        return nil
    }

    private func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method{
        case "gcmEncrypt":
            gcmEncrypt(call, result: result)
        case "gcmDecrypt":
            gcmDecrypt(call, result: result)
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    private func create(_ call: FlutterMethodCall, result: FlutterResult) {
        result(nil)
    }

    private func gcmEncrypt(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        let args = call.arguments as! [String: Any]
        let data = args["data"] as? FlutterStandardTypedData
        let key = args["key"] as? FlutterStandardTypedData
        let nonceSize = args["nonceSize"] as? Int ?? 0

        qryptoWorkItem = DispatchWorkItem {
            var error: NSError?
            let cipherText = CryptoGCMEncrypt(data?.data, key?.data, nonceSize, &error)
            if (error != nil) {
                self.resultError(result: result, error: error)
                return
            }
            
            self.resultSuccess(result: result, resp: cipherText)
        }
        qryptoQueue.async(execute: qryptoWorkItem!)
    }
    
    private func gcmDecrypt(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        let args = call.arguments as! [String: Any]
        let data = args["data"] as? FlutterStandardTypedData
        let key = args["key"] as? FlutterStandardTypedData
        let nonceSize = args["nonceSize"] as? Int ?? 0

        qryptoWorkItem = DispatchWorkItem {
            var error: NSError?
            let plainText = CryptoGCMDecrypt(data?.data, key?.data, nonceSize, &error)
            if (error != nil) {
                self.resultError(result: result, error: error)
                return
            }
            
            self.resultSuccess(result: result, resp: plainText)
        }
        qryptoQueue.async(execute: qryptoWorkItem!)
    }
    
}

File: android/app/src/main/kotlin/org/nkn/mobile/app/crypto/Crypto.kt
package org.nkn.mobile.app.crypto

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import io.flutter.embedding.engine.FlutterEngine
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import org.nkn.sdk.IChannelHandler

class Crypto : IChannelHandler, MethodChannel.MethodCallHandler, EventChannel.StreamHandler, ViewModel() {
    companion object {
        const val METHOD_CHANNEL_NAME = "org.nkn.mobile/native/crypto_method"
        const val EVENT_CHANNEL_NAME = "org.nkn.mobile/native/crypto_event"

        fun register(flutterEngine: FlutterEngine) {
            Crypto().install(flutterEngine.dartExecutor.binaryMessenger)
        }
    }

    lateinit var channel: MethodChannel
    var eventSink: EventChannel.EventSink? = null

    override fun install(binaryMessenger: BinaryMessenger) {
        channel = MethodChannel(binaryMessenger, METHOD_CHANNEL_NAME)
        channel.setMethodCallHandler(this)
    }

    override fun uninstall() {
        channel.setMethodCallHandler(null)
    }

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        eventSink = events
    }

    override fun onCancel(arguments: Any?) {
        eventSink = null
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "gcmEncrypt" -> {
                gcmEncrypt(call, result)
            }
            "gcmDecrypt" -> {
                gcmDecrypt(call, result)
            }

            else -> {
                result.notImplemented()
            }
        }
    }

    private fun gcmEncrypt(call: MethodCall, result: MethodChannel.Result) {
        val data = call.argument<ByteArray>("data")!!
        val key = call.argument<ByteArray>("key")!!
        val nonceSize = call.argument<Int>("nonceSize") ?: 0

        viewModelScope.launch(Dispatchers.IO) {
            try {
                val cipherText = crypto.Crypto.gcmEncrypt(data, key, nonceSize.toLong())
                resultSuccess(result, cipherText)
                return@launch
            } catch (e: Throwable) {
                resultError(result, e)
                return@launch
            }
        }
    }

    private fun gcmDecrypt(call: MethodCall, result: MethodChannel.Result) {
        val data = call.argument<ByteArray>("data")!!
        val key = call.argument<ByteArray>("key")!!
        val nonceSize = call.argument<Int>("nonceSize") ?: 0

        viewModelScope.launch(Dispatchers.IO) {
            try {
                val plainTex = crypto.Crypto.gcmDecrypt(data, key, nonceSize.toLong())
                resultSuccess(result, plainTex)
                return@launch
            } catch (e: Throwable) {
                resultError(result, e)
                return@launch
            }
        }
    }

}